pubsub/gossipsub/gossipsub-v1.1.md

```diff
@@ -0,0 +1,405 @@
+    - [Topic Parameter Calculation and Decay](#topic-parameter-calculation-and-decay)
+      - [P₁: Time in Mesh](#p%E2%82%81-time-in-mesh)
+      - [P₂: First Message Deliveries](#p%E2%82%82-first-message-deliveries)
+      - [P₃ and P₃b: Mesh Message Delivery](#p%E2%82%83-and-p%E2%82%83b-mesh-message-delivery)
+      - [P₄: Invalid Messages](#p%E2%82%84-invalid-messages)
+      - [Parameter Decay](#parameter-decay)
+    - [Guidelines for Tuning the Scoring Function](#guidelines-for-tuning-the-scoring-function)
+  - [Spam Protection Measures](#spam-protection-measures)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+This document specifies extensions to [gossipsub v1.0](gossipsub-v1.0.md) intended to improve
+bootstrapping and protocol attack resistant. The extensions change the algorithms that
+prescribe local peer behaviour and are fully backwards compatible with v1.0 of the protocol.
+Peers that implement these extensions, advertise v1.1 of the protocol using `/meshsub/1.1.0`
+as the protocol string.
+
+## Attack Vectors considered
+
+`To be written`
+
+## Protocol extensions 
+
+### Peer Exchange
+
+Gossipsub relies on ambient peer discovery in order to find peers within a topic of interest.
+This puts pressure to the implementation of a scalable peer discovery service that
+can support the protocol. With Peer Exchange, the protocol can now bootstrap from a small
+set of nodes, without relying on an external peer discovery service.
+
+Peer Exchange (PX) kicks in when pruning a mesh because of oversubscription. Instead of simply
+telling the pruned peer to go away, the pruning peer _may_ provide a set of other peers where the
+pruned peer can connect to reform its mesh. In addition, both the pruned and the pruning peer
+add a backoff period from each other, within which they will not try to regraft. Both the pruning
+and the pruned peer will immediate prune a `GRAFT` within the backoff period.
+The recommended duration for the back period is 1 minute.
+
+In order to implement PX, we extend the `PRUNE` control message to include an optional set of
+peers the pruned peer can connect to. This set of includes the Peer ID and a [_signed_ peer
+record](https://github.com/libp2p/specs/pull/217) for each peer exchanged.
+In order to facilitate transion to the usage of signed peer records within the libp2p ecosystem,
+the emitting peer is allowed to omit the signed peer record if it doesn't have one.
+In this case, the pruned peer will have to utilize an external service to discover addresses for
+the peer, eg the DHT.
+
+#### Protobuf
+
+The `ControlPrune` message is extended with a `peer` field as follows.
+
+```protobuf
+message ControlPrune {
+	optional string topicID = 1;
+	repeated PeerInfo peers = 2; // gossipsub v1.1 PX
+}
+
+message PeerInfo {
+	optional bytes peerID = 1;
+	optional bytes signedPeerRecord = 2;
+}
+
+```
+
+### Flood Publishing
```

Comment:
@vyzo this extension reads more as a consideration than a specification. 

When should a Peer decide to use mesh/fanout or flood? Is there a randomizer for the strategy used?

Reply:
It's an application specification option. The recommendation is that it's used, but the application may turn it off if running in a controlled environment.